### PR TITLE
Escaping html in the chunk

### PR DESCRIPTION
**Pull request by helfi92**
Before:
page-shot-2016-10-17 unified logviewer

After:
screen shot 2016-10-17 at 8 55 30 pm

Tested with URL: [https://archive.mozilla.org/pub/firefox/try-builds/rthijssen@mozilla.com-be36d2322604780a7126392e69be2c8caca37c07/try-win32/try-win32-bm76-try1-build2832.txt.gz
](url)

### CODE REVIEW

File: src/LogViewer/chunk-worker.js [hunk 1]
```diff
@@ -10,6 +10,22 @@ const LINE_CHUNK = 1000;
 const DECODER = new TextDecoder('utf-8');
 // Setting a hard limit on lines since browser have trouble with heights starting at around 16.7 million pixels and up
 const CHUNK_LIMIT = 883000 / LINE_CHUNK;
+// HTML Escape helper utility
```
Comment (eliperelman) [inline src/LogViewer/chunk-worker.js]: I know that this code was taken pretty much as-is from that article I sent you, but I'm going to make some changes and explain.

File: src/LogViewer/chunk-worker.js [hunk 2]
```diff
@@ -10,6 +10,22 @@ const LINE_CHUNK = 1000;
...
+const util = (() => {
```
Comment (eliperelman) [inline src/LogViewer/chunk-worker.js]: I think this makes the escaping a little more abstract than necessary. I think we can change it like so to simplify:

``` js
const ESCAPE_REGEX = /[&<>'"]/g;
const ESCAPE_FUNCTION = entity => ESCAPE_ENTITY[entity];
const ESCAPE_ENTITY = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  "'": '&#39;',
  '"': '&quot;',
  '`': '&#x60;'
};
```

Comment (eliperelman) [reply src/LogViewer/chunk-worker.js]: Also note that I added a new escape for backticks.

File: src/LogViewer/chunk-worker.js [hunk 3]
```diff
@@ -10,6 +10,22 @@ const LINE_CHUNK = 1000;
...
+  const oEscape = {
+    '&': '&amp;',
+    '<': '&lt;',
+    '>': '&gt;',
+    "'": '&#39;',
+    '"': '&quot;'
+  };
+  const fnEscape = (m) => oEscape[m];
+
+  return (Object.freeze || Object)({
```
Comment (eliperelman) [inline src/LogViewer/chunk-worker.js]: With this now simplified into a few constants, we need a function for doing the raw escape:

``` js
const escape = string => string.replace(ESCAPE_REGEX, ESCAPE_FUNCTION);
```

File: src/LogViewer/chunk-worker.js [hunk 4]
```diff
@@ -127,6 +143,18 @@ const getParagraphClass = (lineNumber, { highlightStart, highlightEnd }) => {
     '';
 };
 
+// Tagged template function
```
Comment (eliperelman) [inline src/LogViewer/chunk-worker.js]: We don't need this comment.

File: src/LogViewer/chunk-worker.js [hunk 5]
```diff
@@ -127,6 +143,18 @@ const getParagraphClass = (lineNumber, { highlightStart, highlightEnd }) => {
...
+function escapeHtml(pieces) {
```
Comment (eliperelman) [inline src/LogViewer/chunk-worker.js]: We don't use function declarations in this repo, so let's switch to expressions.

Comment (eliperelman) [reply src/LogViewer/chunk-worker.js]: In looking at this function, I notice a few things that can be cleaned with ES2015:
1. Use of `arguments`. This is almost never needed with ES2015 since we have rest parameters.
2. Variable extraction from parameters, which destructuring and rest parameters also solves.
3. Use of `forEach` to accumulate a value is an indicator that it can be simplified by using `Array#reduce`.

Putting these together, `escapeHtml` can be rewritten like so:

``` js
const escapeHtml = ([result, ...strings], ...values) => values
  .reduce((result, value, index) => `${result}${escape(value)}${strings[index]}`, result);
```

A tagged template receives an array of strings and arguments for the values to be interpolated in between the strings. I renamed the variables to reflect that, so let's walk over the changes:

``` js
function escapeHtml(pieces) {
  let result = pieces[0];
  const substitutions = [].slice.call(arguments, 1);

// becomes

const escapeHtml = ([result, ...strings],
```

The first argument to `escapeHtml` are the strings, so I destructured that into two variables: one for the initial `result`, and the other containing the rest of the `strings`. The remaining arguments are the `substitutions`, which I renamed to `values`.

``` js
substitutions.forEach((elem, i) => {
  result += util.escape(elem) + pieces[i + 1];
});

return result;

// becomes

values.reduce((result, value, index) => `${result}${escape(value)}${strings[index]}`, result);
```
- Iterate over the values: `substitutions.forEach` vs. `values.reduce`
- Start with the current result: `result +=` vs. `.reduce(/**/, result)`
- Concatenate the previous result to the escaped value: `result += util.escape(elem)` vs. `${result}${escape(value)}`
- Concatenate the next piece: `+ pieces[i + 1]` vs. `${strings[index]}`

So the operations are the same, but a little more functional, succinct, and simplified. Also notice that `pieces[i + 1]` is different from `strings[index]` in that `pieces` didn't modify the original array, whereas `strings` was already destructured into an array with the `result` removed.

To recap, this is what the final `escapeHtml` function looks like:

``` js
const escapeHtml = ([result, ...strings], ...values) => values
  .reduce((result, value, index) => `${result}${escape(value)}${strings[index]}`, result);
```

File: src/LogViewer/chunk-worker.js [hunk 6]
```diff
@@ -139,8 +167,8 @@ const toHtml = (index, metadata) => {
       const className = getAnsiClasses(part);
       
       return className ?
-        `<span class="${className}">${part.text}</span>` :
-        `<span>${part.text}</span>`;
+        escapeHtml`<span class="${className}">${part.text}</span>` :
```
Comment (eliperelman) [inline src/LogViewer/chunk-worker.js]: I didn't investigate why this didn't break, but I think only the contents of the tag need to be escaped, since the `span` is part of the HTML that is being rendered to the logviewer. This should be:

``` js
const html = escapeHtml`${part.text}`;

return className ?
  `<span class="${className}">${html}</span>` :
  `<span>${html}</span>`;
```

**Review by eliperelman — CHANGES_REQUESTED** [review]

File: src/LogViewer/chunk-worker.js (hunk 5 above)
Comment (helfi92) [reply src/LogViewer/chunk-worker.js]: @eliperelman This is a gem. I will follow these guidelines for future PRs as well. Thanks!

**Comment by helfi92** [comment]
Added the requested changes

File: src/LogViewer/chunk-worker.js (hunk 5 above)
Comment (eliperelman) [reply src/LogViewer/chunk-worker.js]: @helfi92 my pleasure. A key takeaway is that any usage of `.forEach` is a [code smell](https://en.wikipedia.org/wiki/Code_smell), indicating that the iteration relies on manipulating state instead of following a more functional approach. A `.forEach` can be used more functionally with `.map`, `.reduce`, or `.filter`, but not always. Feel free to ask me if you need advice.

**Review by eliperelman — APPROVED** [review]
Excellent!
